src/jit/inline.cpp

```diff
@@ -474,3 +642,179 @@ void InlineResult::report()
+{
+    // For now, always create a Legacy policy.
+    InlinePolicy* policy = new (compiler, CMK_Inlining) LegacyPolicy;
+
+    return policy;
+}
+
+//------------------------------------------------------------------------
+// noteCandidate: handle passing a set of inlining checks successfully
+//
+// Arguments:
+//    obs      - the current obsevation
+
+void LegacyPolicy::noteCandidate(InlineObservation obs)
+{
+    assert(!inlDecisionIsDecided(inlDecision));
+
+    // Check the impact, it should be INFORMATION
+    InlineImpact impact = inlGetImpact(obs);
+    assert(impact == InlineImpact::INFORMATION);
+
+    // Update the status
+    setCommon(InlineDecision::CANDIDATE, obs);
+}
+
+//------------------------------------------------------------------------
+// noteSuccess: handle finishing all the inlining checks successfully
+
+void LegacyPolicy::noteSuccess()
+{
+    assert(inlDecisionIsCandidate(inlDecision));
+    inlDecision = InlineDecision::SUCCESS;
+}
+
+//------------------------------------------------------------------------
+// note: handle an observation with non-fatal impact
+//
+// Arguments:
+//    obs      - the current obsevation
+
+void LegacyPolicy::note(InlineObservation obs)
+{
+    // Check the impact
+    InlineImpact impact = inlGetImpact(obs);
+
+    // As a safeguard, all fatal impact must be
+    // reported via noteFatal.
+    assert(impact != InlineImpact::FATAL);
+    noteInternal(obs, impact);
+}
+
+//------------------------------------------------------------------------
+// noteFatal: handle an observation with fatal impact
+//
+// Arguments:
+//    obs      - the current obsevation
+
+void LegacyPolicy::noteFatal(InlineObservation obs)
+{
+    // Check the impact
+    InlineImpact impact = inlGetImpact(obs);
+
+    // As a safeguard, all fatal impact must be
+    // reported via noteFatal.
+    assert(impact == InlineImpact::FATAL);
+    noteInternal(obs, impact);
+    assert(inlDecisionIsFailure(inlDecision));
+}
+
+//------------------------------------------------------------------------
+// noteInt: handle an observed integer value
+//
+// Arguments:
+//    obs      - the current obsevation
+//    value    - the value being observed
+
+void LegacyPolicy::noteInt(InlineObservation obs, int value)
+{
+    (void) value;
+    note(obs);
+}
+
+//------------------------------------------------------------------------
+// noteDouble: handle an observed double value
+//
+// Arguments:
+//    obs      - the current obsevation
+//    value    - the value being observed
+
+void LegacyPolicy::noteDouble(InlineObservation obs, double value)
+{
+    (void) value;
+    note(obs);
+}
+
+//------------------------------------------------------------------------
+// setNever: helper for handling an observation
+//
+// Arguments:
+//    obs      - the current obsevation
+//    impact   - impact of the current observation
+
+void LegacyPolicy::noteInternal(InlineObservation obs, InlineImpact impact)
```

Comment:
One of the goals of this stage of refactoring is to move way from having inlining policy be distributed throughout the code base, and instead centralize it within the InlinePolicy class. 

So for example, instead of some importer method making an observation (say, we're reading IL for inlining and we see a switch opcode) and then immediately making a policy determination (this inline must fail), the method will make an observation and forward it to the policy, which will then determine what the observation means.

To get there I first need to make a home for policy decisions (this change), and then start migrating policy code there (subsequent changes).

As part of that migration, observation impacts will slowly start to change away from `FATAL` to less imperative values, and the observation sites will be updated (where suitable) to support policies that don't immediately want to bail out and stop the inline attempt.